libs/autotagging-commons/src/main/java/org/opensearch/rule/service/RulePersistenceService.java

```diff
@@ -0,0 +1,215 @@
+     * @param id - The id of the rule to get.
+     * @param attributeFilters - A map containing the attributes that user want to filter on
+     * @param searchAfter - The sort values from the last document of the previous page, used for pagination
+     * @param listener - ActionListener for GetRuleResponse
+     */
+    public void getRule(
+        String id,
+        Map<Attribute, Set<String>> attributeFilters,
+        String searchAfter,
+        ActionListener<? extends GetRuleResponse> listener
+    ) {
+        // Stash the current thread context when interacting with system index to perform
+        // operations as the system itself, bypassing authorization checks. This ensures that
+        // actions within this block are trusted and executed with system-level privileges.
+        try (ThreadContext.StoredContext context = getContext()) {
+            BoolQueryBuilder boolQuery = buildGetRuleQuery(id, attributeFilters);
+            SearchRequestBuilder searchRequest = client.prepareSearch(RULES_INDEX)
+                .setQuery(boolQuery)
+                .setSize(getMaxReturnSizeAllowedPerGetRequest());
+            if (searchAfter != null) {
+                searchRequest.addSort(_ID_STRING, SortOrder.ASC).searchAfter(new Object[] { searchAfter });
+            }
+            searchRequest.execute(ActionListener.wrap(searchResponse -> handleGetRuleResponse(id, searchResponse, listener), e -> {
+                logger.error("Failed to fetch all rules: {}", e.getMessage());
+                listener.onFailure(e);
+            }));
+        }
+    }
+
+    /**
+     * Builds a bool query to retrieve rules from the rules index, applying attribute-based filters
+     * when needed and ensuring that the rules are associated with the query group feature type.
+     * @param id - The ID of the rule to fetch. If not null, the search will return only this specific rule.
+     * @param attributeFilters - A map of attributes to their associated set of values used to filter the rules.
+     */
+    public BoolQueryBuilder buildGetRuleQuery(String id, Map<Attribute, Set<String>> attributeFilters) {
+        BoolQueryBuilder boolQuery = QueryBuilders.boolQuery();
+        if (id != null) {
+            return boolQuery.must(QueryBuilders.termQuery(_ID_STRING, id));
+        }
+        for (Map.Entry<Attribute, Set<String>> entry : attributeFilters.entrySet()) {
+            Attribute attribute = entry.getKey();
+            Set<String> values = entry.getValue();
+            if (values != null && !values.isEmpty()) {
+                BoolQueryBuilder attributeQuery = QueryBuilders.boolQuery();
+                for (String value : values) {
+                    attributeQuery.should(QueryBuilders.matchQuery(attribute.getName(), value));
+                }
+                boolQuery.must(attributeQuery);
+            }
+        }
+        boolQuery.filter(QueryBuilders.existsQuery(retrieveFeatureTypeInstance().getName()));
```

Comment:
Same goes for this one. My storage layer for the Rules could very well be a sql database hence the conversion of user supplied params will change with underlying storage mechanism.